cpp/src/ray/util/process_helper.cc

```diff
@@ -44,28 +79,51 @@ void ProcessHelper::StopRayNode() {
 
 void ProcessHelper::RayStart(CoreWorkerOptions::TaskExecutionCallback callback) {
   std::string redis_ip = ConfigInternal::Instance().redis_ip;
-  if ((ray::WorkerType)ConfigInternal::Instance().worker_type ==
-          ray::WorkerType::DRIVER &&
+  if (ConfigInternal::Instance().worker_type == ray::WorkerType::DRIVER &&
       redis_ip.empty()) {
     redis_ip = "127.0.0.1";
     StartRayNode(ConfigInternal::Instance().redis_port,
-                 ConfigInternal::Instance().redis_password,
-                 ConfigInternal::Instance().node_manager_port);
+                 ConfigInternal::Instance().redis_password);
+  }
+  if (redis_ip == "127.0.0.1") {
+    redis_ip = GetNodeIpAddress();
   }
 
+  std::string redis_address =
+      redis_ip + ":" + std::to_string(ConfigInternal::Instance().redis_port);
+  std::string node_ip = ConfigInternal::Instance().node_ip_address;
+  if (node_ip.empty()) {
+    if (!ConfigInternal::Instance().redis_ip.empty()) {
+      node_ip = GetNodeIpAddress(redis_address);
+    } else {
+      node_ip = GetNodeIpAddress();
```

Comment:
`GetNodeIpAddress` will be called twice when `redis_ip == "127.0.0.1"` and `node_ip.empty()`?

Reply:
Yes. The logic is the same in Python. See https://github.com/ray-project/ray/blob/bf3e3225b62e813c04982ba72b9ff8d41c62f6e8/python/ray/_private/services.py#L394 https://github.com/ray-project/ray/blob/bf3e3225b62e813c04982ba72b9ff8d41c62f6e8/python/ray/_private/services.py#L429 https://github.com/ray-project/ray/blob/bf3e3225b62e813c04982ba72b9ff8d41c62f6e8/python/ray/node.py#L105